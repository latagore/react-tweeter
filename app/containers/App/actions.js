/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_TWEETS,
  LOAD_TWEETS_SUCCESS,
  LOAD_TWEETS_ERROR,
  CREATE_TWEET,
  CREATE_TWEET_SUCCESS,
  CREATE_TWEET_ERROR
} from './constants';

/**
 * Load the tweets, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_TWEETS
 */
export function createTweet() {
  return {
    type: CREATE_TWEET,
  };
}

/**
 * Dispatched when the tweet is created and the new tweets are loaded
 *
 * @param  {array} tweets The tweet data
 *
 * @return {object}      An action object with a type of CREATE_TWEET_SUCCESS passing the tweets
 */
export function tweetCreated(tweets) {
  return {
    type: CREATE_TWEET_SUCCESS,
    tweets,
  };
}

/**
 * Dispatched when creating the tweet fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of CREATE_TWEET_ERROR passing the error
 */
export function tweetCreatingError(error) {
  return {
    type: CREATE_TWEET_ERROR,
    error,
  };
}


/**
 * Load the tweets, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_TWEETS
 */
export function loadTweets() {
  return {
    type: LOAD_TWEETS,
  };
}

/**
 * Dispatched when the tweets are loaded by the request saga
 *
 * @param  {array} tweets The tweet data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_TWEETS_SUCCESS passing the tweets
 */
export function tweetsLoaded(tweets) {
  return {
    type: LOAD_TWEETS_SUCCESS,
    tweets,
  };
}

/**
 * Dispatched when loading the tweets fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_TWEETS_ERROR passing the error
 */
export function tweetsLoadingError(error) {
  return {
    type: LOAD_TWEETS_ERROR,
    error,
  };
}
