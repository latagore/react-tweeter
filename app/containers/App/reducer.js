/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_TWEETS_SUCCESS,
  LOAD_TWEETS,
  LOAD_TWEETS_ERROR,
  CREATE_TWEET,
  CREATE_TWEET_SUCCESS,
  CREATE_TWEET_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  tweets: [],
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_TWEETS:
      return state
        .set('loading', true)
        .set('error', false);
    case LOAD_TWEETS_SUCCESS:
      return state
        .set('tweets', action.tweets)
        .set('loading', false);
    case LOAD_TWEETS_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    case CREATE_TWEET:
      return state
        .set('loading', true)
        .set('error', false);
    case CREATE_TWEET_SUCCESS:
      let newState = state;
      if (action.tweets) {
        newState = newState
          .set('tweets', action.tweets);
      }
      return newState
        .set('loading', false);
    case CREATE_TWEET_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}

export default appReducer;
