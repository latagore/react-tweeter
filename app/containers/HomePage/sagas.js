/**
 * Gets tweets from the server
 */

import { take, call, put, select, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { LOAD_TWEETS } from 'containers/App/constants';
import { tweetsLoaded, tweetsLoadingError } from 'containers/App/actions';

import request from 'utils/request';

/**
 * Github repos request/response handler
 */
export function* getTweets() {
  // fake data to test the app
  // const fakeTweets = [
  //   {
  //     user: {
  //       handle: 'Dog',
  //       avatarImageURL: 'http://i.imgur.com/Iek1A.png',
  //     },
  //     content: 'Hello world!',
  //   },
  // ];
  const requestURL = `http://localhost:3000/api/feed`;

  try {
    const tweets = yield call(request, requestURL);
    yield put(tweetsLoaded(tweets));
  } catch (err) {
    yield put(tweetsLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* tweetData() {
  // Watches for LOAD_TWEETS actions and calls getTweets when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_TWEETS, getTweets);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  tweetData,
];
