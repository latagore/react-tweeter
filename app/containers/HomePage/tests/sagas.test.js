/**
 * Tests for HomePage sagas
 */

import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';

import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_REPOS } from 'containers/App/constants';
import { tweetsLoaded, tweetsLoadingError } from 'containers/App/actions';

import { getTweets, githubData } from '../sagas';

/* eslint-disable redux-saga/yield-effects */
describe('getTweets Saga', () => {
  let getTweetsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getTweetsGenerator = getTweets();

    // const selectDescriptor = getTweetsGenerator.next().value;
    // expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getTweetsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the tweetsLoaded action if it requests the data successfully', () => {
    const response = [{
      name: 'First tweet',
    }, {
      name: 'Second tweet',
    }];
    const putDescriptor = getTweetsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(tweetsLoaded(response)));
  });

  it('should call the tweetsLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getTweetsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(tweetsLoadingError(response)));
  });
});
//
// describe('githubDataSaga Saga', () => {
//   const githubDataSaga = githubData();
//   const mockedTask = createMockTask();
//
//   it('should start task to watch for LOAD_REPOS action', () => {
//     const takeLatestDescriptor = githubDataSaga.next().value;
//     expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_REPOS, getTweets));
//   });
//
//   it('should yield until LOCATION_CHANGE action', () => {
//     const takeDescriptor = githubDataSaga.next(mockedTask).value;
//     expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
//   });
//
//   it('should cancel the forked task when LOCATION_CHANGE happens', () => {
//     const cancelDescriptor = githubDataSaga.next().value;
//     expect(cancelDescriptor).toEqual(cancel(mockedTask));
//   });
// });
