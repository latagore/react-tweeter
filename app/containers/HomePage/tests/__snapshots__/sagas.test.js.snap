// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTweets Saga should call the tweetsLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "localhost:3000/api/feed",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getTweets Saga should dispatch the tweetsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "localhost:3000/api/feed",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
